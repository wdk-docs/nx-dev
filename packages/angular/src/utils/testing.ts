import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Rule, Tree } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/angular',
  join(__dirname, '../../generators.json')
);

testRunner.registerCollection(
  '@nrwl/jest',
  join(__dirname, '../../../jest/generators.json')
);

testRunner.registerCollection(
  '@nrwl/workspace',
  join(__dirname, '../../../workspace/generators.json')
);

testRunner.registerCollection(
  '@nrwl/cypress',
  join(__dirname, '../../../cypress/generators.json')
);

testRunner.registerCollection(
  '@nrwl/storybook',
  join(__dirname, '../../../storybook/generators.json')
);

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/workspace',
  join(__dirname, '../../migrations.json')
);

export function runMigration<SchemaOptions = any>(
  schematicName: string,
  options: SchemaOptions,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync(schematicName, options, tree)
    .toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}
