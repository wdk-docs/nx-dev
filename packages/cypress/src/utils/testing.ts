import { join } from 'path';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { Tree, Rule } from '@angular-devkit/schematics';

const testRunner = new SchematicTestRunner(
  '@nrwl/cypress',
  join(__dirname, '../../generators.json')
);

const migrationRunner = new SchematicTestRunner(
  '@nrwl/cypress/migrations',
  join(__dirname, '../../migrations.json')
);

export function runSchematic(schematicName: string, options: any, tree: Tree) {
  return testRunner.runSchematicAsync(schematicName, options, tree).toPromise();
}

export function callRule(rule: Rule, tree: Tree) {
  return testRunner.callRule(rule, tree).toPromise();
}

export function runMigration(migrationName: string, options: any, tree: Tree) {
  return migrationRunner
    .runSchematicAsync(migrationName, options, tree)
    .toPromise();
}
