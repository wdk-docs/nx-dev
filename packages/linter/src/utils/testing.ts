import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import { join } from 'path';
import { Tree } from '@angular-devkit/schematics';

const migrationTestRunner = new SchematicTestRunner(
  '@nrwl/linter-migrations',
  join(__dirname, '../../migrations.json')
);

export function runMigration<T = any>(
  migrationName: string,
  options: T,
  tree: Tree
) {
  return migrationTestRunner
    .runSchematicAsync<T>(migrationName, options, tree)
    .toPromise();
}
