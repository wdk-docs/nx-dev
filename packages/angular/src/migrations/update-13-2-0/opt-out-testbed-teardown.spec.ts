import { addProjectConfiguration, logger, Tree } from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import testbedTeardownOptOutMigration from './opt-out-testbed-teardown';

describe('opt-out-testbed-teardown migration', () => {
  let tree: Tree;

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace(2);
    jest.doMock('@nrwl/tao/src/utils/app-root', () => ({ appRootPath: '' }));
  });

  it('should warn when the jestConfig property is not configured', async () => {
    // Arrange
    jest.spyOn(logger, 'warn');
    addProjectConfiguration(tree, 'app1', {
      root: 'apps/app1',
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {},
        },
      },
    });

    // Act
    await testbedTeardownOptOutMigration(tree);

    // Assert
    expect(logger.warn).toHaveBeenCalledWith(
      'The "jestConfig" property is not configured for the test target of the project "app1". Skipping it.'
    );
  });

  it('should warn when the specified jestConfig file does not exist', async () => {
    // Arrange
    jest.spyOn(logger, 'warn');
    addProjectConfiguration(tree, 'app1', {
      root: 'apps/app1',
      targets: {
        test: {
          executor: '@nrwl/jest:jest',
          options: {
            jestConfig: 'some-path/jest.config.js',
          },
        },
      },
    });

    // Act
    await testbedTeardownOptOutMigration(tree);

    // Assert
    expect(logger.warn).toHaveBeenCalledWith(
      'The specified "jestConfig" path "some-path/jest.config.js" for the project "app1" can not be found. Skipping it.'
    );
  });

  describe('update initTestEnvironment calls in test setup', () => {
    it('should update correctly when there are no calls to initTestEnvironment', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        'import "jest-preset-angular/setup-jest";'
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when no options are passed to the initTestEnvironment call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting());
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when the options passed to the initTestEnvironment call is an empty object', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), {});
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when the options passed to the initTestEnvironment call does not have the teardown property', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { aotSummaries: () => [] });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when an aotSummaries function is passed to the initTestEnvironment call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), function() { return []; });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when an aotSummaries named function is passed to the initTestEnvironment call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), function aotSummaries() { return []; });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should update correctly when an aotSummaries arrow function is passed to the initTestEnvironment call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), () => []);
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });

    it('should not re-add the teardown property twice when it is already being passed to the initTestEnvironment call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { teardown: { destroyAfterEach: true } });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testSetupFile = tree.read('apps/app1/src/test-setup.ts', 'utf-8');
      expect(testSetupFile).toMatchSnapshot();
    });
  });

  describe('update configureTestingModule in test files', () => {
    it('should not be updated when the test setup file was updated', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        'import "jest-preset-angular/setup-jest";'
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should not be updated when the test setup file had the teardown options already configured', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { teardown: { destroyAfterEach: true } });
        `
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should update correctly when teardown options and setupFilesAfterEnv are not configured', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should update correctly when teardown options have not been configured and there is no test setup file configured', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: [],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should update correctly when teardown options have not been configured and the specified test setup file does not exist', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should update correctly when teardown options have not been configured and the initTestEnvironment call is invalid', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          setupFilesAfterEnv: ['<rootDir>src/test-setup.ts'],
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/test-setup.ts',
        `import "jest-preset-angular/setup-jest";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        // this is a type error, missing 2nd parameter
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule);
        `
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });`
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should update correctly when there are multiple calls to configureTestingModule', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent1', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });

        describe('AppComponent2', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });

        describe('AppComponent3', () => {
          test('some test case', async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          
            expect(true).toBe(true);
          });
        });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });

    it('should not re-add the teardown property or overwrite when it is already configured for configureTestingModule call', async () => {
      // Arrange
      const jestConfigPath = 'apps/app1/jest.config.js';
      addProjectConfiguration(tree, 'app1', {
        root: 'apps/app1',
        targets: {
          test: {
            executor: '@nrwl/jest:jest',
            options: { jestConfig: jestConfigPath },
          },
        },
      });
      tree.write(
        jestConfigPath,
        `module.exports = {
          transform: { '^.+.(ts|mjs|js|html)$': 'jest-preset-angular' },
        };`
      );
      tree.write(
        'apps/app1/src/app.component.spec.ts',
        `describe('AppComponent1', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });

        describe('AppComponent2', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
              teardown: { destroyAfterEach: true },
            }).compileComponents();
          });
        });

        describe('AppComponent3', () => {
          test('some test case', async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          
            expect(true).toBe(true);
          });
        });
        `
      );

      // Act
      await testbedTeardownOptOutMigration(tree);

      // Assert
      const testFile = tree.read(
        'apps/app1/src/app.component.spec.ts',
        'utf-8'
      );
      expect(testFile).toMatchSnapshot();
    });
  });
});
