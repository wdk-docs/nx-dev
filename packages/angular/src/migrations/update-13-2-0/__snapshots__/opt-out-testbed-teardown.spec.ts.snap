// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should not be updated when the test setup file had the teardown options already configured 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should not be updated when the test setup file was updated 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
            }).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should not re-add the teardown property or overwrite when it is already configured for configureTestingModule call 1`] = `
"describe('AppComponent1', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });

        describe('AppComponent2', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
              declarations: [AppComponent],
              teardown: { destroyAfterEach: true },
            }).compileComponents();
          });
        });

        describe('AppComponent3', () => {
          test('some test case', async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          
            expect(true).toBe(true);
          });
        });
        "
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should update correctly when teardown options and setupFilesAfterEnv are not configured 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should update correctly when teardown options have not been configured and the initTestEnvironment call is invalid 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should update correctly when teardown options have not been configured and the specified test setup file does not exist 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should update correctly when teardown options have not been configured and there is no test setup file configured 1`] = `
"describe('AppComponent', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });"
`;

exports[`opt-out-testbed-teardown migration update configureTestingModule in test files should update correctly when there are multiple calls to configureTestingModule 1`] = `
"describe('AppComponent1', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });

        describe('AppComponent2', () => {
          beforeEach(async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          });
        });

        describe('AppComponent3', () => {
          test('some test case', async () => {
            await TestBed.configureTestingModule({
    declarations: [AppComponent],
    teardown: { destroyAfterEach: false }
}).compileComponents();
          
            expect(true).toBe(true);
          });
        });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should not re-add the teardown property twice when it is already being passed to the initTestEnvironment call 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { teardown: { destroyAfterEach: true } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when an aotSummaries arrow function is passed to the initTestEnvironment call 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { aotSummaries: () => [], teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when an aotSummaries function is passed to the initTestEnvironment call 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { aotSummaries: function () { return []; }, teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when an aotSummaries named function is passed to the initTestEnvironment call 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { aotSummaries: function aotSummaries() { return []; }, teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when no options are passed to the initTestEnvironment call 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when the options passed to the initTestEnvironment call does not have the teardown property 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { aotSummaries: () => [], teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when the options passed to the initTestEnvironment call is an empty object 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
        import { getTestBed } from '@angular/core/testing';
        import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

        getTestBed().resetTestEnvironment();
        getTestBed().initTestEnvironment(BrowserDynamicTestingModule, platformBrowserDynamicTesting(), { teardown: { destroyAfterEach: false } });
        "
`;

exports[`opt-out-testbed-teardown migration update initTestEnvironment calls in test setup should update correctly when there are no calls to initTestEnvironment 1`] = `
"import \\"jest-preset-angular/setup-jest\\";
      import { getTestBed } from '@angular/core/testing';
      import { BrowserDynamicTestingModule, platformBrowserDynamicTesting } from '@angular/platform-browser-dynamic/testing';

      getTestBed().resetTestEnvironment();
      getTestBed().initTestEnvironment(
        BrowserDynamicTestingModule,
        platformBrowserDynamicTesting(),
        { teardown: { destroyAfterEach: false } },
      );
    "
`;
